File: Rust/src/solution/my_queue.rs

pub struct MyQueue {
    stack_in: Vec<i32>,
    stack_out: Vec<i32>,
}

impl MyQueue {

    pub fn new() -> Self {
        MyQueue {
            stack_in: Vec::new(),
            stack_out: Vec::new(),
        }
    }

    pub fn push(&mut self, x: i32) {
        self.stack_in.push(x);
    }

    pub fn pop(&mut self) -> i32 {
        self.move_to_out();
        self.stack_out.pop().unwrap()
    }

    pub fn peek(&mut self) -> i32 {
        self.move_to_out();
        *self.stack_out.last().unwrap()
    }

    pub fn empty(&self) -> bool {
        self.stack_in.is_empty() && self.stack_out.is_empty()
    }

    fn move_to_out(&mut self) {
        if self.stack_out.is_empty() {
            while let Some(val) = self.stack_in.pop() {
                self.stack_out.push(val);
            }
        }
    }
}

//  * Your MyQueue object will be instantiated and called as such:
//  * let obj = MyQueue::new();
//  * obj.push(x);
//  * let ret_2: i32 = obj.pop();
//  * let ret_3: i32 = obj.peek();
//  * let ret_4: bool = obj.empty();

File: Cs/Solution/MyQueue.cs

using System.Collections.Generic;

namespace Solution
{

	public class MyQueue
	{

		private Stack<int> stack1;
		private Stack<int> stack2;

		public MyQueue()
		{
			stack1 = new Stack<int>();
			stack2 = new Stack<int>();
		}

		public void Push(int x)
		{
			stack1.Push(x);
		}

		public int Pop()
		{
			TransferStack1ToStack2();
			return stack2.Pop();
		}

		public int Peek()
		{
			TransferStack1ToStack2();
			return stack2.Peek();
		}

		public bool Empty()
		{
			return (stack1.Count == 0) && (stack2.Count == 0);
		}

		private void TransferStack1ToStack2()
		{
			if (stack2.Count == 0)
			{
				while (stack1.Count > 0)
				{
					stack2.Push(stack1.Pop());
				}
			}
		}
	}
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 * int param_3 = obj.Peek();
 * bool param_4 = obj.Empty();
 */
